Update deployment doc (#130)
atolab/zenoh-web: content/docs/migration_1.9/Regions.md

---
title: "Regions"
weight : 7100
menu:
  docs:
    parent: migration_1.9
---

## Zenoh regions

> Zenoh version 1.9.0 introduces a new feature that allows subdividing a system into _regions_ in a more flexible manner than before. This provides more scalability and allows deployments and topologies that were not possible before.

Reviewer: Suggestion: add a link to the 1.9.0 blog post here.

Author: Done

---
title: "Regions"
weight : 7100
menu:
  docs:
    parent: migration_1.9
---

## Zenoh regions

> Zenoh version 1.9.0 introduces a new feature that allows to subdivise a system in regions in a more flexible manner than before. This provides more scalability and allows deployments and topologies that were not possible before.

Reviewer: ```suggestion
Zenoh version 1.9.0 introduces a new feature that allows subdividing a system into _regions_ in a more flexible manner than before. This provides more scalability and allows deployments and topologies that were not possible before.
```

### The Zenoh Gateway

> Version 1.9.0 introduces the Zenoh *Gateway*. A Zenoh gateway is a Zenoh node that can serve one or several sub regions called *south* regions. It will act as a representant of the south region(s) in the main region called the *north* region.

Reviewer: ```suggestion
Version 1.9.0 introduces the Zenoh *Gateway*. A Zenoh gateway is a Zenoh node that can serve one or several subregions called *south* regions. It will act as a representative of the south region(s) in the main region called the *north* region.
```

- Use "subregion", or even "sub-region" to eliminate ambiguity: I first read this as "subscriber region."
- "Representant" is [obsolete](https://en.wiktionary.org/wiki/representant). I would also go for "proxy."

> The `mode` of the Zenoh gateway node defines what communication model is used in the north region and thus how the node should act in the north region. A Zenoh gateway can accept any communication models in it's south regions and so do not need any configuration of south regions modes.

Reviewer: ```suggestion
The `mode` of the Zenoh gateway node defines what communication model is used in the north region and thus how the node should act in the north region. A Zenoh gateway can accept any communication model in each of its south regions. Thus configuration of south regions' modes is not necessary.
```

![gateay](/img/migration_1.9/gateway.png "gateway")

Each region will use one of the three currently available communication patterns:

**Routed:**
In routed regions, Zenoh nodes must be deployed with `router` mode and can be deployed in a mesh topology.

![routed](/img/migration_1.9/routed.png "routed")

**Peer to Peer:**

> In peer to peer regions, Zenoh nodes must be deployed with `peer` mode and must be deployed in a clique topology where every node is connected to all other nodes of the region.

Reviewer: ```suggestion
In peer to peer regions, Zenoh nodes must be deployed with `peer` mode and must also be deployed in a clique topology where every node is connected to all other nodes of the region.
```

![peer2peer](/img/migration_1.9/peer2peer.png "peer2peer")

**Brokered:**

> In brokered regions, Zenoh nodes must be deployed with `client` mode and must be connected to a "broker" node. The broker node is typically a gateway to a north region or the super node of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.

Reviewer: ```suggestion
In brokered regions, Zenoh nodes must be deployed with `client` mode and must also be connected to a "broker" node. The broker node is typically a gateway to a north region or the super node of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.
```

> In brokered regions, Zenoh nodes must be deployed with `client` mode and must be connected to a "broker" node. The broker node is typically a gateway to a north region or the super node of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.

Reviewer: ```suggestion
In brokered regions, Zenoh nodes must be deployed with `client` mode and must be connected to a "broker" node. The broker node is typically a gateway to a north region or one of the "northernmost" nodes of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.
```

Also consider scrapping the sentence entirely. I don't think that brokers are either gateways to upstream regions or _the (only) northernmost node_ in the system. They can have be peers or routers in their north region and so they can be many.

> In brokered regions, Zenoh nodes must be deployed with `client` mode and must be connected to a "broker" node. The broker node is typically a gateway to a north region or the super node of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.

Reviewer: ```suggestion
In brokered regions, Zenoh nodes must be deployed with `client` mode and must be connected to a "broker" node. The broker node is typically a gateway to a north region or one of the "northernmost" nodes of the system. The mode of the broker node is irrelevant for the brokered region and only affects the north region.
```

Also consider scrapping the sentence entirely. I don't think that brokers are either gateways to upstream regions or _the (only) northernmost node_ in the system. They can have be peers or routers in their north region and so they can be many.

![brokered](/img/migration_1.9/brokered.png "brokered")

#### Scalability considerations

> The gateway will hide non needed details of the sub region(s) to the upper region (number of nodes, topology, individual subscribers and queryables, etc..). It will also hide non needed details of the upper region to the sub regions. This provides more scalability to Zenoh systems.

Reviewer: ```suggestion
The gateway will hide non needed details of the subregion(s) to its north region (number of nodes, topology, individual subscribers and queryables, etc.) It will also hide unecessary details of the north region from the subregions. This provides more scalability to Zenoh systems.
```

### Limitations

#### Regions hierarchy

> While different topologies and communication models are possible inside each Zenoh region, the Zenoh regions **must** be deployed in a **hierarchical** manner. A Zenoh region can have at most one north region. 

Reviewer: ```suggestion
While different topologies and communication models are possible inside each Zenoh region, Zenoh regions **must** be deployed in a **hierarchical** manner. A Zenoh region can have at most one north region. 
```

> Note: for load balancing and fault tolerance purposes, multiple gateways may be deployed to serve a same south region in a north region, but those multiple gateways **must** be deployed in the same north reion.

Reviewer: ```suggestion
Note: for load balancing and fault tolerance purposes, multiple gateways may be deployed to serve one south region, but those multiple gateways **must** be deployed in the same north region.
```

![hierarchy](/img/migration_1.9/hierarchy.png "hierarchy")

#### Routed regions

Routed regions (regions in which Zenoh nodes are deployed with the `router` mode and in which Zenoh nodes can be deployed in a mesh topology) can only be deployed in the south of other routed regions. 

> Routed regions **cannot** be deployed in the south of a peer-to-peer region or in the south of a brokered region.

Reviewer: ```suggestion
In particular, routed regions **cannot** be deployed in the south of a peer-to-peer region or in the south of a brokered region.
```

### Configuration

> Zenoh gateways are configured by specifying through filters which connected nodes should go to the north region and which connected nodes should go to the south regions (and which south region). The connected nodes can be filtered by modes, zids, network interface and/or region names. 

Reviewer: ```suggestion
Zenoh gateways are configured through filters that specify which connected nodes should go to the north region and which connected nodes should go to a given south region. The connected nodes can be filtered by any combination of modes, zids, network interfaces and region names. 
```

> This configuration is located in the `gateway/south` section of the Zenoh configuration. The default value of this section is: `"auto"`. When set to `"auto"`, Zenoh will filter nodes according to their mode in the most natural manner and similarly to previous versions. Routers will always go to the north while non gateway clients will go to the south.

Reviewer: ```suggestion
This configuration is located in the `gateway/south` section of the Zenoh configuration. The default value of this section is: `"auto"`. When set to `"auto"`, Zenoh will filter nodes according in such a manner as to maintain the behavior of prior versions. For example, `"auto"` routers will always go to the north of other `"auto"` nodes, while `"auto"` clients will always go to the south of other `"auto"` nodes.
```

I don't think that "natural manner" would be clear to readers, better scrap it.

Example:
```
{
  gateway: {
    south: "auto",
  }
}
```

> For more advanced gateways configurations, the `gateway/south` section should contain a list of south regions. Each region should containing a list of filters. All nodes matching one of the filters will be affected to the corresponding south region.

Reviewer: ```suggestion
For more advanced gateways configurations, the `gateway/south` section should contain a list of south regions. Each region should contain a list of filters. All nodes matching one of the filters will be affected to the corresponding south region.
```

Each filter can contain a list of conditions: `modes`, `zids`, `interfaces` and/or `region_names`. Each condition is a list of possible values. A filter can also have an optional `negate` boolean. 
- A node matches a condition, if it matches one of the possible values. 

> - A node matches a filter if 

Reviewer: ```suggestion
- A node matches a filter if:
```

    - it matches all the conditions of the filter. 

>     - or if `negate` is `true` and the node deos not match one of the conditions of the filter.

Reviewer: ```suggestion
    - or if `negate` is `true` and the node does not match one of the conditions of the filter.
```

- A node matches a south region if it matches one of the filters of the region. 

Example of configuration with two south regions:
```
{
  gateway: {
    south: [
      {
        filters: [
          {
            modes: ["peer", "client"],
            zids: ["aa", "bb", "cc"]
          }
        ]
      },
      {
        filters: [
          {
            interfaces: ["wlan0"],
            negate: true
          }
        ]
      }
    ]
  }
}
```

> Nodes that do not match any south region and has the correct mode for the north region will be affected to the north region.

Reviewer: ```suggestion
Nodes that do not match any south region and have the correct mode for the north region will be affected to the north region.
```

> Nodes that do not match any south region and don't have the correct mode for the north region will be affected to the *unbound* region. The  *unbound* region behaves like a south region.

Reviewer: This is not true as of 1.9.0: such nodes will simply fail to connect.

#### Region names

A region name can be assigned to Zenoh nodes.

Example:
```
{
  region_name: "region1"
}
```

> Nodes can be filtered on their region name in the `gateway` configuration.

Reviewer: ```suggestion
Nodes can be filtered by their region name in the `gateway` configuration.
```

Example of configuragtion for nodes `A` and `B`:
```
{
  mode: "client",
  listen: {
    endpoints: ["tcp/[::]:0"]
  },
  connect: {
    endpoints: ["tcp/router_ip:7447"]
  },
  gateway: {
    south: [
      {
        filters: [
          {
            modes: ["client", "peer"]
          }
        ]
      }
    ]
  }
}
```

Example of configuragtion for nodes `C`, `D`, `E` and `F`:
```
{
  mode: "peer"
}
```

#### South regions interconnected in peer to peer

![topology2](/img/migration_1.9/new_topo2.png "topology2")

Example of configuragtion for nodes `A` and `B`:
```
{
  mode: "peer",
  region_name: "main",
  gateway: {
    south: [
      {
        filters: [
          {
            negated: true,
            region_names: ["main"]
          }
        ]
      }
    ]
  }
}
```

Example of configuragtion for nodes `C`, `D`, `E` and `F`:
```
{
  mode: "peer"
}
```

## Migration

Most scenarios and topologies will continue working without any change when migrating from Zenoh version 1.8.x to 1.9.0. But:
- Some configuration option defaults have been changed.
- Some features and configuration options have been removed or changed in Zenoh v1.9.0.

> ### Changed defaults

Reviewer: This would be more readable as a table with three columns: option, previous default, new default.

- `connect.timeout_ms` default value changed 
  - from `{ router: -1, peer: -1, client: 0 }` 
  - to `-1`
- `connect.exit_on_failure` default value changed 
  - from `{ router: false, peer: false, client: true }` 
  - to `false`
- `scouting.timeout` default value changed 
  - from `3000` 
  - to `-1`
- `scouting.multicast.listen` default value changed
  - from `{ router: true, peer: true, client: false }` 
  - to `true`
- `scouting.multicast.autoconnect` default value changed
  - from `{ router: [], peer: ["router", "peer"], client: ["router"] }`
  - to `{ router: [], peer: ["router", "peer", "client"], client: ["router", "peer", "client"] }`
- `scouting.gossip.autoconnect` default value changed
  - from `{ router: [], peer: ["router", "peer"], client: ["router"] }`
  - to `{ router: [], peer: ["router", "peer", "client"], client: ["router", "peer", "client"] }`

### Removed options

Here is a list of configuration options that have been removed in Zenoh v1.9.0 and instructions on how to change the Zenoh configuration to deploy similar scenarios with Zenoh v1.9.0. 

#### `peers_failover_brokering` option

> The `routing/router/peers_failover_brokering` feature has been removed and the related option is no longer available in the configuration. This feature allowed the following kind of topology:

Reviewer: ```suggestion
The peer failover brokering feature in routers has been removed and the corresponding option `routing/router/peers_failover_brokering` is no longer available in the configuration. This feature allowed the following kind of topology:
```

Here `routing/router/peers_failover_brokering` refers to the option, but it is called the feature. Then the "related option" is mentioned right after.

![topology1](/img/migration_1.9/migration_topo1.png "topology1")

This kind of deployment will not work any more out of the box.

Here are 2 ways to workaround this change:

> **Deploy one gateway per sub region**

Reviewer: ```suggestion
**Deploy one gateway per subregion**
```
